vpr/src/route/rr_graph_generation/rr_node_indices.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include "rr_graph_builder.h"
+#include "rr_graph_view.h"
+#include "device_grid.h"
+#include "rr_types.h"
+#include "rr_graph_type.h"
+#include "rr_graph_utils.h"
+#include "clustered_netlist_fwd.h"
+
+/**
+ * @brief Allocates and populates data structures for efficient rr_node index lookups.
+ *
+ * This function sets up the `rr_node_indices` structure, which maps a physical location
+ * and type to the index of the first corresponding rr_node.
+ */
+void alloc_and_load_rr_node_indices(RRGraphBuilder& rr_graph_builder,
+                                    const t_chan_width& nodes_per_chan,
+                                    const DeviceGrid& grid,
+                                    int* index,
+                                    const t_chan_details& chan_details_x,
+                                    const t_chan_details& chan_details_y);
+
+/**
+ * @brief Allocates extra nodes within the RR graph to support 3D custom switch blocks for multi-die FPGAs
+ *
+ *  @param rr_graph_builder RRGraphBuilder data structure which allows data modification on a routing resource graph
+ *  @param nodes_per_chan number of tracks per channel (x, y)
+ *  @param grid device grid
+ *  @param extra_nodes_per_switchblock keeps how many extra length-0 CHANX node is required for each unique (x,y) location within the grid.
+ *  Number of these extra nodes are exactly the same for all layers. Hence, we only keep it for one layer. ([0..grid.width-1][0..grid.height-1)
+ *  @param index RRNodeId that should be assigned to add a new RR node to the RR graph
+ */
+void alloc_and_load_inter_die_rr_node_indices(RRGraphBuilder& rr_graph_builder,
+                                              const t_chan_width& nodes_per_chan,
+                                              const DeviceGrid& grid,
+                                              const vtr::NdMatrix<int, 2>& extra_nodes_per_switchblock,
+                                              int* index);
+
+void alloc_and_load_tile_rr_node_indices(RRGraphBuilder& rr_graph_builder,
```

Comment:
If you know what the two functions below do this would be a good time to comment them.